### ISSUE TITLE
Is there cases in which minTempo and maxTempo are not expected to be respected?

### ISSUE THREAD

**Issue by mrgreenh** [chrisguttandin/web-audio-beat-detector#362]
Hi @chrisguttandin , first of all thank you so much for your work on this library, it's a real lifesaver for working with music and javascript and I've really been enjoying using it so far.

I have recently hit an issue with the usage of minTempo and maxTempo and I was hoping to shed some light on what might be going on. I have noticed that if I run this track named "Flanger Danger by Poli, 3 Of Life" ([on my dropbox](https://www.dropbox.com/scl/fi/km78wi6wfjufrkulfjt5d/Flanger_Danger-Poli-_3_Of_Life-Original_Mix.mp3?rlkey=p6yvz2mhiz40w5z9sw1vp71xl&dl=0) for your convenience) through `guess(audioBuffer, { maxTempo: 155, minTempo: 120 })`, even though the track's bpm is 142, I get a result of "114".

I don't have a full understanding of the mechanism behind this library, but based on what I found [here](https://github.com/chrisguttandin/web-audio-beat-detector-worker/blob/master/src/helpers/group-neighbors-by-tempo.ts#L13-L18) I'm reaching the conclusion that the tempo range I provide is only used to adjust for the cases in which the detected tempo is half or double the correct one (e.g. when a 140bpm track is detected to be 70bpm). Is this correct? And if yes, do you know if there is any ideas to be considered to improve on this? I have been using [bpm-tools](https://www.pogo.org.uk/~mark/bpm-tools/) in the past and these appear to always return results between the specified range, so I wonder if there is some solution to be borrowed.

Thank you in advance for your time and reply!

**Comment by chrisguttandin** [comment]
Hi @mrgreenh,

thanks for reporting this. This is actually a very tricky problem. The piece of code you linked assumes that the `maxTempo` value is twice the `minTempo` value. My intenion definitely was to make sure the result is within the given range. But dividing the value by 2 or multiplying it by 2 may result in a value outside of the given range. I'm not sure what the library should do when that happens.

Besides all that there seems to be something going wrong when analyzing your example file.

**Comment by mrgreenh** [comment]
Hi @chrisguttandin thank you for the clarification. That's interesting, I tried setting maxTempo to 240 now and it does indeed  result in 227. However I agree it's an odd behavior because it's as if it's not leveraging the range as a further hint for running the statistical analysis, so that if the result was wrong (by that I mean the tempo is just completely off, not just multiplied by the wrong multiple of 2) without a provided range, it's still going to be wrong when a range is provided.

Based on what you are describing, it might make sense to update the API to allowing for specifying only one of the 2 bounds, something like `{bound: number, type: 'lower | upper'}` to set the correct expectations.

(Short of having the analysis behave the same as bpm-tools which however I realize might be a whole different kind of challenge :D )

**Comment by mrgreenh** [comment]
Quick followup, I completely randomly stumbled into this: in an instance in which I accidentally input the min and max bounds *inverted* (so I accidentally set maxTempo to 120 and minTempo to 155), the `guess()` call for that same audio file spit out... 71!!! Which is half of 142 and thus a more correct and useful result than 114.... Just thought this hint might help figure out what's going on.

**Comment by chrisguttandin** [comment]
Thanks a lot for providing further details. I think this is also a symptom of the same bug. If a range is provided it will influence the weights of the tempo candidates which will in turn produce a different result. I changed the code to ignore results that are outside of the expected range.

The results for the test file are now as follows:

```js
guess(audioBuffer); // bpm: 114
guess(audioBuffer, { maxTempo: 155, minTempo: 120 }); // bpm: 142
guess(audioBuffer, { maxTempo: 120, minTempo: 155 }); // no result
```

The first result is probably because the algorithm picks up some off beats. The second result is correct. And I would argue the last one makes sense, too. There are no results in the given range which means it can't find any.

I hope this all makes sense.

The changes got released as v8.2.11.
